File: crates/topcoat-runtime/browser/src/expr/expr_deref.ts
import { type Expr, eval_expr } from "./index";
import type { Interpreter } from "./interpreter";

export type ExprDeref = {
	type: "Deref";
	inner: Expr;
};

export function eval_expr_deref(
	expr: ExprDeref,
	interpreter: Interpreter,
): unknown {
	const target = eval_expr(expr.inner, interpreter);
	return (target as () => unknown)();
}

File: crates/topcoat-runtime/src/expr/expr_deref.rs
use serde::{Serialize, Serializer, ser::SerializeStruct};

use crate::{Expr, Interpreter};

pub trait ExprDerefTarget {
    type Target;

    fn expr_deref(self) -> Self::Target;
}

pub struct ExprDeref<E>(E);

impl<E> ExprDeref<E> {
    pub fn new(inner: E) -> Self {
        Self(inner)
    }
}

impl<E> Expr for ExprDeref<E>
where
    E: Expr,
    E::Output: ExprDerefTarget,
{
    type Output = <E::Output as ExprDerefTarget>::Target;

    fn eval(self, interpreter: &mut Interpreter) -> Self::Output {
        self.0.eval(interpreter).expr_deref()
    }
}

impl<E: Serialize> Serialize for ExprDeref<E> {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        let mut s = serializer.serialize_struct("ExprDeref", 2)?;
        s.serialize_field("type", "Deref")?;
        s.serialize_field("inner", &self.0)?;
        s.end()
    }
}
